src/wallet/walletutil.cpp

```diff
@@ -57,35 +57,60 @@ bool IsBerkeleyBtree(const fs::path& path)
 std::vector<fs::path> ListWalletDir()
 {
     const fs::path wallet_dir = GetWalletDir();
+    const fs::path data_dir = GetDataDir();
+    const fs::path blocks_dir = GetBlocksDir();
+
     const size_t offset = wallet_dir.string().size() + 1;
     std::vector<fs::path> paths;
     boost::system::error_code ec;
 
+    // Here we place the top level dirs we want to skip in case walletdir is datadir or blocksdir
+    // Those directories are referenced in doc/files.md
+    const std::vector<fs::path> ignore_paths = {
+                                        blocks_dir,
+                                        data_dir / "blktree",
+                                        data_dir / "chainstate",
+                                        data_dir / "coins",
+                                        data_dir / "database",
+                                        data_dir / "indexes",
+                                        data_dir / "regtest",
+                                        data_dir / "signet",
+                                        data_dir / "testnet3"
+                                        };
+
     for (auto it = fs::recursive_directory_iterator(wallet_dir, ec); it != fs::recursive_directory_iterator(); it.increment(ec)) {
         if (ec) {
             LogPrintf("%s: %s %s\n", __func__, ec.message(), it->path().string());
             continue;
         }
 
-        // Get wallet path relative to walletdir by removing walletdir from the wallet path.
-        // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
-        const fs::path path = it->path().string().substr(offset);
-
-        if (it->status().type() == fs::directory_file && IsBerkeleyBtree(it->path() / "wallet.dat")) {
-            // Found a directory which contains wallet.dat btree file, add it as a wallet.
-            paths.emplace_back(path);
-        } else if (it.level() == 0 && it->symlink_status().type() == fs::regular_file && IsBerkeleyBtree(it->path())) {
-            if (it->path().filename() == "wallet.dat") {
-                // Found top-level wallet.dat btree file, add top level directory ""
-                // as a wallet.
-                paths.emplace_back();
-            } else {
-                // Found top-level btree file not called wallet.dat. Current bitcoin
-                // software will never create these files but will allow them to be
-                // opened in a shared database environment for backwards compatibility.
-                // Add it to the list of available wallets.
-                paths.emplace_back(path);
+        // We don't want to iterate through those special node dirs
+        if (std::count(ignore_paths.begin(), ignore_paths.end(), it->path())) {
```

Comment:
nit: [std::find](https://en.cppreference.com/w/cpp/algorithm/find) could return earlier.

Reply:
yup. no insist here from my side, but nit too if "not"  *const time* could that sidechannel reveal where the wallet is or else?

Btw> since u review here and in  #18095, 18095 does fix some problems, but not the wallet of death, could we transfer attention off reviewers to this here and fix some bugs fast in one sweep, before adding there new ones?
I must admit my PR headlines lock often a bit harmless and just like minor nits, but when I way in and insist, the hut is burning. just saying .

Reply:
I don't think that `std::count` _guaranties_ const time to prevent sidechannel reveal where the wallet is or else. But I'm not an expert in this field. Feel free to mark this comment as resolved.

Reply:
Ok, then i resolve this here and fix the header. thx for review.